coldfront/core/department/utils/ldap.py

```diff
@@ -0,0 +1,115 @@
+    """Search for a user in LDAP.
+
+    Parameters:
+        email (str): Email address of user
+        first_name (str): First name of user
+        last_name (str): Last name of user
+    Returns:
+        (ldap3.Entry) LDAP entry for user or None if not found
+    """
+    conn = Connection(LDAP_URL, auto_bind=True)
+    if conn.search(PEOPLE_OU, f'(&(objectClass=person)(mail={email}))',
+                                            attributes=['departmentNumber']):
+        return conn.entries[0]
+    elif conn.search(PEOPLE_OU, '(&(objectClass=person)'
+            f'(givenName={first_name})(sn={last_name}))',
+            attributes=['departmentNumber']) and len(conn.entries) == 1:
+        return conn.entries[0]
+    return None
+
+
+def get_L4_code_from_department_code(code):
+    """Return the L4 code for a department from a L4+ code.
+
+    Parameters:
+        code (str): L4+ department code
+    Returns:
+        (str) Normalized L4 department code
+    """
+    conn = Connection(LDAP_URL, auto_bind=True)
+    conn.search(DEPARTMENT_OU,
+                f'(&(objectClass=organizationalUnit)(ou={code}))',
+                attributes=['berkeleyEduOrgUnitHierarchyString'])
+    hierarchy = conn.entries[0].berkeleyEduOrgUnitHierarchyString.value
+    return hierarchy.split('-')[3]
+
+
+def get_department_name_from_code(code):
+    """Return the name of a department from a code.
+
+    Parameters:
+       code (str): department code
+    Returns:
+        (str) Name of department
+    """
+    conn = Connection(LDAP_URL, auto_bind=True)
+    conn.search(DEPARTMENT_OU,
+                f'(&(objectClass=organizationalUnit)(ou={code}))',
+                attributes=['description'])
+    return conn.entries[0].description.value
+
+
+def fetch_and_set_user_departments(user, userprofile, dry_run=False):
+    """Fetch a user's departments from LDAP and sets them in the
+    database.
+
+    Parameters:
+        user (User): Django User object
+        userprofile (UserProfile): Coldfront UserProfile object
+        dry_run (bool): If True, don't actually make changes
+                        (default: False)
+    Returns:
+        None
+    """
+    if dry_run:
+        logger.info = print
+        department = Department()
+        department.pk = '{placeholder_pk}'
+        userdepartment = UserDepartment()
+        userdepartment.pk = '{placeholder_pk}'
+    user_entry = ldap_search_user(user.email, user.first_name, user.last_name)
+    ldap_departments = user_entry.departmentNumber.values if user_entry else []
+    for department_code in ldap_departments:
+        department_code = get_L4_code_from_department_code(department_code)
+        # If a Department doesn't exist, create it.
+        name = None if Department.objects.filter(code=department_code).exists() \
+            else get_department_name_from_code(department_code)
+        if dry_run:
+            created = name is not None
+            department.code = department_code
```

Comment:
if dry_run is true, it's defined at the beginning of the function. otherwise, it gets defined on line 95. It's not referenced when dry_run is false until after line 95. Is there another way it could be referenced before assignment that I'm missing?

Reply:
I agree that it won't happen, just responding to my IDE's warning.

You could place the code in the first `if dry_run` block into the `for` loop.

![warning](https://github.com/ucb-rit/coldfront/assets/10734545/50846580-6397-4294-a312-2ebb844dc432)